### PR TITLE
Add source & stage decorators

### LINKED ISSUE #1253: [FEA]: Create a stage decorator allowing stages to be created from Python functions
### Is this a new feature, an improvement, or a change to existing functionality?

New Feature

### How would you describe the priority of this feature request

Medium

### Please provide a clear description of problem this feature solves

For stages requiring only a single on-data method, a stage author should be able to do something like:
```
@stage(name="pass-thru")
def pass_thru(msg: MessageMeta) -> MessageMeta:
    return msg
```

### Describe your ideal solution

Make a decorator which removes the need to define a new class.

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct
- [X] I have searched the [open feature requests](https://github.com/nv-morpheus/Morpheus/issues?q=is%3Aopen+is%3Aissue+label%3A%22feature+request%22%2Cimprovement%2Cenhancement) and have found no duplicates for this feature request

### PR DESCRIPTION
**Pull request by dagardner-nv**
## Description
* Adds two new decorators `source` and `stage` allowing stages to be created from python functions.
* Accept & return types are inferred by inspecting the type annotations of the decorated function.
* This removes the need for several lines of boilerplate code needed for simple Python-only stages.
* Includes unrelated docstring fix to `morpheus/pipeline/stage_base.py`

example:
```py
from morpheus.pipeline.stage_decorator import source
from morpheus.pipeline.stage_decorator import stage

@source
def source_gen(dataframes: list[cudf.DataFrame]) -> collections.abc.Iterator[MessageMeta]:
    for df in dataframes:
        yield MessageMeta(df)


@stage
def on_data(message: MessageMeta) -> MessageMeta:
    with message.mutable_dataframe() as df:
        print(df)
    return message
```
 
Closes #1253

## By Submitting this PR I confirm:
- I am familiar with the [Contributing Guidelines](https://github.com/nv-morpheus/Morpheus/blob/main/docs/source/developer_guide/contributing.md).
- When the PR is ready for review, new or existing tests cover these changes.
- When the PR is ready for review, the documentation is up to date with these changes.

### CODE REVIEW

**Review by drobison00 — APPROVED** [review]

**Review by mdemoret-nv — CHANGES_REQUESTED** [review]
This is a good start but we need some changes to be forward looking with these decorators. General changes:

1. The classes used by the decorators should be "dumber" and move most of the implementation into the decorator instead of the class. See my comments for more details.
2. Need some tests validating the parameters used at construction time
3. Needs documentation showing how this can be used in the developer guide.

File: morpheus/pipeline/stage_decorator.py [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+    # annotations to reflect that the returned function requires a config and returns a stage
+    @functools.wraps(on_data_fn, assigned=('__module__', '__name__', '__qualname__', '__doc__'))
+    def wrapper(config: Config, *args, **kwargs) -> WrappedFunctionStage:
+        return WrappedFunctionStage(*args,
+                                    config=config,
+                                    on_data_fn=on_data_fn,
+                                    needed_columns=needed_columns,
+                                    **kwargs)
+
+    return wrapper
```
[the remark this answers is not in the document]
Comment (dagardner-nv) [reply morpheus/pipeline/stage_decorator.py:338]: We do currently support arguments to the stage decorator, it works as both `@stage` and `@stage()`

**Comment by dagardner-nv** [comment]
> 3. Needs documentation showing how this can be used in the developer guide.

I do have dev-doc updates, I just did them in a different branch. I can merge it back in here.

File: morpheus/pipeline/stage_decorator.py [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+    ...
+    ...     return message
+    ...
+    >>>
+
+    >>> pipe.add_stage(multiplier(config, column='v2', value=5))
+    """
+
+    if on_data_fn is None:
+        return functools.partial(stage, needed_columns=needed_columns)
```
[the remark this answers is not in the document]
Comment (dagardner-nv) [reply morpheus/pipeline/stage_decorator.py:326]: That allows us to support both `@stage` and `@stage()` invocations.

File: morpheus/pipeline/stage_decorator.py [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+        Return type of `gen_fn` if not provided the stage will use the return type annotation of `gen_fn` as the output.
+    *on_data_args : `typing.Any`
+        Additional arguments to bind to `on_data_fn` via `functools.partial`.
+    **on_data_kwargs : `typing.Any`
+        Additional keyword arguments to bind to `on_data_fn` via `functools.partial`.
+    """
+
+    def __init__(self,
+                 config: Config,
+                 on_data_fn: typing.Callable,
```
[the remark this answers is not in the document]
Comment (dagardner-nv) [reply morpheus/pipeline/stage_decorator.py]: I think we should still allow `config` to be positional as it is specified positional in the documentation & examples for all other stages :
```
    def __init__(self,
                 config: Config,
                 *,
                 on_data_fn: typing.Callable,
```

[the remark this answers is not in the document]
Comment (mdemoret-nv) [reply morpheus/pipeline/stage_decorator.py]: Thats fine.

**Review by mdemoret-nv — APPROVED** [review]
Looks good now. Like the changes. This will be super useful in the future.

**Pull request merged**